public/tracy/TracyMetal.hmm

```diff
@@ -0,0 +1,634 @@
+            reallocateBuffer = true;
+        }
+        ZoneValue(count);
+        
+        auto buffer_idx = (begin / MaxQueries) % 2;
+        auto counterSampleBuffer = m_counterSampleBuffers[buffer_idx];
+
+        if (count >= RingSize())
+        {
+            TracyMetalPanic(return false, "Collect: FULL! too many pending timestamp queries. [%llu, %llu] (%u)", begin, latestCheckpoint, count);
+        }
+
+        TracyMetalDebug(1<<3, TracyMetalPanic(, "Collect: [%llu, %llu] :: (%u)", begin, latestCheckpoint, count));
+
+        NSRange range = NSMakeRange(RingIndex(begin), count);
+        NSData* data = [counterSampleBuffer resolveCounterRange:range];
+        NSUInteger numResolvedTimestamps = data.length / sizeof(MTLCounterResultTimestamp);
+        MTLCounterResultTimestamp* timestamps = (MTLCounterResultTimestamp *)(data.bytes);
+        if (timestamps == nil)
+        {
+            TracyMetalPanic(return false, "Collect: unable to resolve timestamps.");
+        }
+
+        if (numResolvedTimestamps != count)
+        {
+            TracyMetalPanic(, "Collect: numResolvedTimestamps != count : %u != %u", (uint32_t)numResolvedTimestamps, count);
+        }
+
+        int resolved = 0;
+        for (auto i = 0; i < numResolvedTimestamps; i += 2)
+        {
+            ZoneScopedN("TracyMetal::Collect::[i]");
+            MTLTimestamp t_start = timestamps[i+0].timestamp;
+            MTLTimestamp t_end = timestamps[i+1].timestamp;
+            uint32_t k = RingIndex(begin + i);
+            TracyMetalDebug(1<<4, TracyMetalPanic(, "Collect: timestamp[%u] = %llu | timestamp[%u] = %llu | diff = %llu\n", k, t_start, k+1, t_end, (t_end - t_start)));
+            if ((t_start == MTLCounterErrorValue)  || (t_end == MTLCounterErrorValue))
+            {
+                TracyMetalPanic(, "Collect: invalid timestamp (MTLCounterErrorValue) at %u.", k);
+                break;
+            }
+            // Metal will initialize timestamp buffer with zeroes; encountering a zero-value
+            // timestamp means that the timestamp has not been written and resolved yet
+            if ((t_start == 0) || (t_end == 0))
+            {
+                auto checkTime = std::chrono::high_resolution_clock::now();
+                auto requestTime = m_timestampRequestTime[k];
+                auto ms_in_flight = std::chrono::duration<float>(checkTime-requestTime).count()*1000.0f;
+                TracyMetalDebug(1<<4, TracyMetalPanic(, "Collect: invalid timestamp (zero) at %u [%.0fms in flight].", k, ms_in_flight));
+                const float timeout_ms = 200.0f;
+                if (ms_in_flight < timeout_ms)
+                    break;
+                ZoneScopedN("TracyMetal::Collect::Drop");
+                TracyMetalPanic(, "Collect: giving up on timestamp at %u [%.0fms in flight].", k, ms_in_flight);
+                t_start = m_mostRecentTimestamp + 5;
+                t_end = t_start + 5;
```

Comment:
Does this means you try tp "patch" unresolved or not yet resolved timestamps ? 
Can't we defer their resolution instead ?

Reply:
We can't defer, because the main reason a timestamp can't be resolved is because it the command encoder with the associated timestamp(s) was never scheduled for execution.